archive/indexer.go

```diff
@@ -54,59 +83,61 @@ func CreateIndexes(dir string) error {
 	return err
 }
 
-func IndexLogFile(path string) error {
-	subdir := path + zarExt
-	zdxName := "zdx:type:ip"
-	zdxPath := filepath.Join(subdir, zdxName)
-	// XXX remove without warning, should have force flag
-	zdx.Remove(zdxPath)
-
-	fmt.Printf("%s: indexing as %s\n", path, zdxPath)
-
-	file, err := os.Open(path)
-	if err != nil {
-		return err
-	}
-	defer file.Close()
-	// XXX should be able to index any kind of logs... find should take
-	// a regexp to match files and then use auto-detect or -i for input format
-	reader := bzngio.NewReader(file, resolver.NewContext())
-	table, err := indexTypeIP(reader)
+func Run(path string, rules []Rule, progress chan string) error {
+	subdir, err := archiveDir(path)
 	if err != nil {
 		return err
 	}
-	if table.Size() == 0 {
-		//XXX
-		return errors.New("nothing to index")
-	}
-	// make subdirectory for index if it doesn't exist
-	if err := os.Mkdir(subdir, 0755); err != nil {
-		if !os.IsExist(err) {
+	var indexers []Indexer
+	for _, rule := range rules {
+		indexer, err := rule.NewIndexer(subdir)
+		if err != nil {
 			return err
 		}
+		indexers = append(indexers, indexer)
+		if progress != nil {
+			progress <- fmt.Sprintf("%s: creating index %s", path, indexer.Path())
+		}
 	}
-	framesize := 32 * 1024
-	//XXX for now specify value size of 0, which means variable, but we always
-	// write nil values.  we should change the implementation to allow key-only zdx files.
-	writer, err := zdx.NewWriter(zdxPath, framesize, 0)
+	file, err := os.Open(path)
 	if err != nil {
 		return err
 	}
-	defer writer.Close()
-	return zdx.Copy(writer, table)
-}
-
-func indexTypeIP(reader zbuf.Reader) (*zdx.MemTable, error) {
-	table := zdx.NewMemTable()
-	indexer := &TypeIndexer{Type: zng.TypeIP, Table: table}
+	defer file.Close()
+	reader := bzngio.NewReader(file, resolver.NewContext())
+	// XXX This for-loop could be easily parallelized by having each writer
+	// live in its own go routine and sending the rec over a set of
+	// blocking channels (so we flow-control it).
 	for {
 		rec, err := reader.Read()
 		if err != nil {
-			return nil, err
+			return err
 		}
 		if rec == nil {
-			return table, nil
+			break
+		}
+		for _, indexer := range indexers {
+			if err := indexer.Write(rec); err != nil {
+				return err
+			}
+		}
+	}
+	// we make the framesize here larger than the writer framesize
+	// since the writer always writes a bit past the threshold
+	const framesize = 32 * 1024 * 2
```

Comment:
Does this comment still make sense? (I'm not sure what "the writer framesize" refers to.)

Reply:
I'd like to leave this for now.  We should have an issue to fix this up.  The indexer buffer should be  larger than threshold used by the writer because it's a threshold rather than a limit.